### ISSUE TITLE
Rethink (or maybe even remove?) ``super``

### ISSUE THREAD

**Issue by ekpyron** [ethereum/solidity#7261]
Given the inheritance changes in 0.6.0 I think we should check where and how ``super`` fits into that and whether we should keep, remove or change it.
E.g. in
```
contract A { function f() virtual {} }
contract B { function f() virtual {} }
contract C is A B { function f() override(A, B) {
  super.f();
} }
```
using ``super`` would defeat the purpose of the whole construction with overrides and unambiguous inheritance...

**Comment by chriseth** [comment]
With the explicit overrides, the burden of checking that inheritance is correct is shifted to the derived contract, but there are still cases where `super` is useful, and these are the ones where we have a shared base:
```
contract I {
  function set() public virtual {}
}
contract A is I {
  uint a;
  function set() public virtual override { a = 1; super.set(); a = 2; }
}
contract B is I {
  uint b;
  function set() public virtual override { b = 1; super.set(); b = 2; }
}
contract X is A, B {
  function set() public override(A, B) { super.set(); }
}
```

These contracts are still valid according to our new rules and all functions will be called in sequence. Without `super`, the derived contract has to call the base functions, but it will not be possible anymore, to call the super function "in between".

**Comment by chriseth** [comment]
Having said that, not sure if there is a use-case for that :)

**Comment by ekpyron** [comment]
You did mean both ``A is I`` and ``B is I``, right? Yeah, this is a point, but to me this always seemed like a really weird construct...

**Comment by chriseth** [comment]
Yes, edited.

**Comment by Marenz** [comment]
Wouldn't `I.set()` also need `virtual` here?

**Comment by chriseth** [comment]
Yes, edited.

**Comment by ekpyron** [comment]
Maybe we should bring this up in the user group as well to get some input on whether this remaining super construction has actual use cases?

**Comment by ekpyron** [comment]
@cameel commented on gitter: ``Maybe making it not look that "constant" would help?``.

For which examples to consider would be ``super()`` or ``super(this)``.

EDIT: ``super(this)`` would make it look like it will become an external call, though, unfortunately...

**Comment by ekpyron** [comment]
In any case: the existence of issues like https://github.com/ethereum/solidity/issues/8354 also indicates that the meaning of ``super`` isn't entirely clear...

**Comment by cameel** [comment]
After our discussion, I think that the syntax could help a little but won't solve the problem on its own. The feature just has complex behavior that needs to be understood.

The same feature exists in Python and there, in my experience, it's not a big issue only because most of the time the inheritance hierarchy is not that complicated and you don't have to think about it until you run into actual problems.

As for syntax, Python 2 does `super` like this:
```python
class MoleculePropertiesView(MoleculePropertiesViewBase):
    def __init__(self):
        super(MoleculePropertiesView, self).__init__(self)
```

while in Python 3 simplified it to just `super()`:
```python
class MoleculePropertiesView(MoleculePropertiesViewBase):
    def __init__(self):
        super().__init__(self)
```

**Comment by ekpyron** [comment]
Given https://github.com/ethereum/solidity/issues/7744 we could consider making it ``super(type(this))``... that would look slightly less like an external call... but yeah as @cameel just said: maybe changing the syntax won't help in the end and might make things worse or just differently confusing...

**Comment by cameel** [comment]
Going back to the example from the description. I think that we could improve the situation a little by having the documentation explicitly say when **not** to use `super`. Currently it shows that you can use it to solve the problem with the diamond pattern but it may not be clear that it's not something appropriate in all situations. In your example the explicit `A.f(); B.f();` is actually what **should** be used. One could shoehorn `super` into it with something like `A.f(); super.f();` but that's awkward and unnecessary.

**Comment by frangio** [comment]
I agree in this case `super` should not be used. But if you had `B is A` and B itself did use super, then using `super.f()` is the only correct way to implement `C.f()`.

By this I mean to say that super should not be removed.

Unless of course it is replaced by some other equivalent mechanism. Like the more explicit notion of before-, after-, and around-methods in CLOS.

**Comment by frangio** [comment]
Is this covered by https://github.com/ethereum/solidity/issues/13727 ?
